autogen/middleware/base.py

```diff
@@ -0,0 +1,189 @@
+    h: MiddlewareCallable = h_async if inspect.iscoroutinefunction(f) else h_sync  # type: ignore[assignment]
+
+    h._origin = f  # type: ignore[misc]
+    h._chained_call = f  # type: ignore[misc]
+
+    # if the name of the first argument is `self` or `cls`, then we assume
+    # this is a method, otherwise it is a function
+    sig = inspect.signature(f)
+    first_param_name = list(sig.parameters.keys())[0]
+    h._is_method = first_param_name in ("self", "cls")  # type: ignore[misc]
+
+    # if async, then we need to await the result
+    h._is_async = inspect.iscoroutinefunction(f)  # type: ignore[misc]
+
+    h._middlewares = []  # type: ignore[misc]
+
+    return h  # type: ignore[return-value]
+
+
+def _get_next_function(h: MiddlewareCallable, mw: Middleware, next: Callable[..., Any]) -> Callable[..., Any]:
+    next = match_caller_type(callee=next, caller=mw.call)
+    call = match_caller_type(callee=mw.call, caller=h._origin)
+    trigger = match_caller_type(callee=mw.trigger, caller=h._origin) if hasattr(mw, "trigger") else None
+
+    if inspect.iscoroutinefunction(h._origin):
+
+        async def _aa() -> None:
+            pass  # pragma: no cover
+
+        next_async = match_caller_type(callee=next, caller=_aa)
+
+        @wraps(call)
+        async def _chain_async(*args: Any, **kwargs: Any) -> Any:
+            trigger_value = (await trigger(*args, **kwargs)) if trigger is not None else True
+            if trigger_value:
+                return await call(*args, next=next, **kwargs)
+            else:
+                return await next_async(*args, **kwargs)
+
+        return _chain_async
+    else:
+
+        def _ss() -> None:
+            pass  # pragma: no cover
+
+        next_sync = match_caller_type(callee=next, caller=_ss)
+
+        @wraps(call)
+        def _chain_sync(*args: Any, **kwargs: Any) -> Any:
+            if trigger is None or trigger(*args, **kwargs):
+                return call(*args, next=next, **kwargs)
+            else:
+                return next_sync(*args, **kwargs)
+
+        return _chain_sync
+
+
+def _build_middleware_chain(h: MiddlewareCallable) -> None:
+    next = h._origin
+
+    for mw in reversed(h._middlewares):
+        next = _get_next_function(h, mw, next)
+        mw._chained_call = next  # type: ignore[attr-defined]
+
+    h._chained_call = next  # type: ignore[misc]
+
+
+def _check_middleware(h: MiddlewareCallable, mw: Middleware) -> None:
+    if inspect.iscoroutinefunction(mw.call) and not inspect.iscoroutinefunction(h):
+        raise TypeError(f"Cannot use middleare with async `call` method on a sync hookable function: {mw=}, {h=} ")
+    if hasattr(mw, "trigger") and inspect.iscoroutinefunction(mw.trigger) and not inspect.iscoroutinefunction(h):
+        raise TypeError(f"Cannot use middleare with async `trigger` method on a sync hookable function: {mw=}, {h=} ")
+
+
+def add_middleware(h: Callable[..., Any], mw: Any, *, position: Optional[int] = None) -> None:
```

Comment:
Instead of modifying original `h` maybe it would be better to create a wrapped_h and return that wrapped_h instead? Otherwise the middleware might effect external scope.

For example
```python
agent: ConversableAgent

def do_something(agent):
    add_middleware(agent.generate_reply, logMiddleware('A'))
   # code...

do_something(agent)
# the logMiddleware('A') still have effect here even though it's already out of scope, which will be very confusing
```

Reply:
If you run the code above, you will modify the `generate_reply` method all current and future instances of `ConversableAgent`, except for those instances that have their `generate_reply` method already modified. This is because in Python functions, methods and classes are all object references. Here is a simpler example:

```python
class MyClass:
    def method(self):
        print("Original Method")


# Create two instances
instance1 = MyClass()
instance2 = MyClass()

# Now both instance1's and instance2's `method` references `MyClass`'s `method`.

# Modify the method of instance1
instance1.method = lambda: print("Modified Method for instance1")

# Now instance1's `method` references the new lambda function, while instance2's `method` still references the original one.

# Call the method on both instances
instance1.method()  # This will print "Modified Method for instance1"
instance2.method()  # This will still print "Original Method"

MyClass.method = lambda _: print("Modified Method for MyClass")

# Now instance1's `method` still references the first lambda function, however, the reference to `MyClass`'s `method` has been updated. So instance2's `method` now references the second lambda function. 

instance1.method()  # This will print "Modified Method for instance1"
instance2.method()  # This will now print "Modified Method for MyClass"


instance3 = MyClass()
instance3.method()  # This will now print "Modified Method for MyClass"

# As shown above, all future instances of `MyClass` will have the updated `method`.
```

In this PR, `add_middleware` simply updates the input function, which is an object by itself, with attributes including the list of middlewares. So if you use it on the class's function, then it is going to affect all instances, except for those instances with their function already referencing something else.

Reply:
I believe all current and future instances of `ConversableAgent` will have their `generate_reply` method using the logging middleware. 

@davorrunje come to think of it, I think we want the effect to localize on instance of agents. Perhaps we want to combine the `register_for_middleware` decorator with `add_middleware`. So a user can simply do:

```python
agent_a = ConversableAgent()

add_middleware(agent_a.generate_reply, LoggingMiddleware())

# Now agent_a has logging capability.

agent_b = ConversableAgent()
# Agent b has no logging capability
```

Should be easy to do. We always make a new Middleware callable object using the input function. If the input function is not a Middleware callable, then we create a new one. If the input function is already a Middleware callable, then we also create a new one with combined middleware.

Reply:
@LittleLittleCloud and @ekzhu you are absolutely right! We need to make middleware specific to an instance and not the class. Constructors should be used to set up the default middleware, but a user should be able to change the behaviour of each instance without affect on other objects.

Reply:
refactored as suggested 👍